src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseSign1Message.cs

```diff
@@ -85,36 +130,75 @@ public static byte[] Sign(byte[] content!!, AsymmetricAlgorithm key!!, HashAlgor
                 writer.WriteByteString(content);
             }
 
-            writer.WriteByteString(signature);
+            int expectedToBeSignedSize = ComputeToBeSignedEncodedSize(SigStructureCoxtextSign1, encodedProtectedHeaders, content);
+
+            Span<byte> toBeSignedBuffer = buffer;
+            byte[]? rentedToBeSignedBuffer = null;
+            int signatureBytesWritten;
+
+            // It is possible for toBeSigned to be bigger than the COSE message length that we used to determine the size of our buffer.
+            // we rent a bigger buffer if that's the case.
+            if (buffer.Length < expectedToBeSignedSize)
+            {
+                rentedToBeSignedBuffer = ArrayPool<byte>.Shared.Rent(expectedToBeSignedSize);
+                toBeSignedBuffer = rentedToBeSignedBuffer;
+            }
+
+            try
+            {
+                int toBeSignedBytesWritten = CreateToBeSigned(SigStructureCoxtextSign1, encodedProtectedHeaders, content, toBeSignedBuffer);
+                ReadOnlySpan<byte> encodedToBeSigned = buffer.Slice(0, toBeSignedBytesWritten);
+
+                if (keyType == KeyType.ECDsa)
+                {
+                    signatureBytesWritten = SignWithECDsa((ECDsa)key, encodedToBeSigned, hashAlgorithm, buffer);
+                }
+                else
+                {
+                    signatureBytesWritten = SignWithRSA((RSA)key, encodedToBeSigned, hashAlgorithm, buffer);
+                }
+            }
+            finally
+            {
+                if (rentedToBeSignedBuffer != null)
+                {
+                    ArrayPool<byte>.Shared.Return(rentedToBeSignedBuffer);
```

Comment:
The only time we would have had to rent the TBS destination is when the content was detached.

Since the caller didn't want their content embedded in the output let's assume it's secret, and clear this buffer before returning it.

```suggestion
                    buffer.Slice(0, toBeSignedBytesWritten).Clear();
                    ArrayPool<byte>.Shared.Return(rentedToBeSignedBuffer);
```

Reply:
Right, this can be performed as
```suggestion
                    ArrayPool<byte>.Shared.Return(rentedToBeSignedBuffer, clearArray: true);
```